### PR TITLE
#9982 Alternative layouts for the TOC

### PR DESCRIPTION
**Pull request by allyoucanmap**
## Description

This PR introduces following changes in the TOC:

- Review of the TOC container system, now all the buttons related to a plugin are injected instead to be hardcoded. Now following target will be available for container injection:
  - `toolbar` add a button in the toolbar
  - `context-menu` add a button to the context menu
  - `node-tool` add a button inside a node
  - `node` replace a node
- Review of the drag and drop system
- Review of the style to make it more compact
- Review of the visibility system. Now groups has their own visibility state that could make inactive (so not visible) the children nodes
- Introduction of the context menu
- Introduction of groups with mutually exclusive nodes
- Removal of FloatingLegend plugin from the core
- Introduction of persisted configuration in the saved map

Preview visibility, expand and context menu:

https://github.com/geosolutions-it/MapStore2/assets/19175505/af2d90ea-df90-4b4b-8dc5-c830404b6fe1

Preview toc settings:

https://github.com/geosolutions-it/MapStore2/assets/19175505/df7aae08-e1d3-453d-af26-93cad9b31efa

Preview dashboard legend and TOC:

https://github.com/geosolutions-it/MapStore2/assets/19175505/bd166cc5-7363-4751-8693-5a9c947f03b6

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Feature

## Issue

**What is the current behavior?**

#9982

**What is the new behavior?**

Refactor of TOC plugin

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [x] Yes, and I documented them in migration notes

## Other useful information

Please check migration guideline to test also the `MetadataInfo` plugin configuration. This plugin read CSW metadata.

### CODE REVIEW

**Review by offtherailz — CHANGES_REQUESTED** [review]
Generally speaking, it is a great job. A code review is quite hard, and if it need some changes, it risks to be also unuseful. So I'll limit this review on the functional aspect or bugs.

I have some suggestion for the general usability. See below

- *note*, I sent some inline comments to the translation files, but they can be ignored if some of the suggestions below are accepted

# Bugs

- [x] **glitch of dnd** I found this issue trying to edit a geostory, but I think it can happen also in other cotnexts. The problems looks to be due to the move or the element that makes the group shift in another position, so the current mouse position is not the same anymore, respectively of the node pointed. The element should not be removed until the drop is done. This may help. Only the target position should be indicated (trying to not move nodes or add extra-space).

[screencast-mail.google.com-2024.03.14-15_17_00.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/0badc228-7b0a-431c-af29-d052faec06a2)

- [x] **ghost drop target** 
The header and the footer of the group are highlighted as droptarget but they are not valid. This confuses about the effective position where things can be dropped.

- [ ] **right click menu doesn't work in geostory media editor** ( IDK if this is by desing, I try to click and it doesn't show anything). 

[screencast-mail.google.com-2024.03.14-15_34_50.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/c4eef645-fbd0-4a7e-9d0f-44c4223ba388)

# Suggestions

## TOC

- [ ]The initial TOC empty view is a little to basic (and what if the catalog is not present ?). I suggest to make the text centered with an icon (using empty view). Possible button from the catalog can be injected. 

**Current**
image

**Suggested** (use your text, "layer tree is empty" is a good sub-text too. I realized it after doing the screenshot :smile:  )

image


- [ ] When selecting the Group, you should somehow highlight a little also the surrounding box, to visually show the group selected, even if you are not selecting the child nodes. Now it gives the sensation of selecting only the title line, not the group. Actions are relative to the group, so it could be better to highlight the group. 

**Current**:
image

**Suggested** (find a better color anyway, not sure if the sub-nodes should have main- background color while only the box is colored):
image




- [ ] Empty groups should contain some space to make drag and drop easier, and to make them to be identified as groups as well. 


**Current**: 

image


**Suggested**: 
image

(css used for the box in the suggested example here)
```css
background-color: rgba(0, 0, 0, .1);
    margin: 5px;
    padding: 5px;
    height: 30px;
    text-align: center;
    border-radius: 2px;
    border: dashed 1px lightgray;
    width: 100%;
```

The bug of the drag-drop may be mitigated by an empty view like this?

**one more observation
- I see the old maps have the default group "locked" and not draggable. There should be the possibility to unlock/update the toc, in order to make easier to update new maps to the new format. Is there a way ?

## Toc settings Menu

image

- [ ] The arrow icon is not clear for "Open on map initialization".  The "open" icon looks to me more intuitive.
image

- [ ] The Menu is hard to understand because of the "opposite" label logic. For instance on the UI you see "Keep closed on map initialization", that means that it actually opens. This is confusing. Most of these options have 2 states. I should use a checkbox or an active/inactive to indicate the current state, and if you want to explain more what happens when you click, a tooltip.

```
- [ ] Open on map load ("initialization" sound strange)
- [ ] Draggable layers (in fact what distinguish the 2 themes is only the draggable layers as far as I can see)
- [ ] Hide overflowing title
- [ ] Truncate long titles
- [ ] Opacity slider
- [ ] Opacity slider's tooltip
```
If you want to preserve the icons, you can put the checkbox on the right, or make the lines look active/not active. 

## Right click menu

- [ ] As every right click option, the menu is not visible or accesible, and for mobile usage can become hard. I see the menu appears on long tap. I suggest to use a "vertical 3 dots button" on the right of both layers and groups. It can be still accessible with right click, but it needs to be accessible also from a visible tool. See here the "Measure" layer and the "Florence" group as example.
image

## Group right click menu

- [ ] Here I'd simplify the logic. 

```
- Show all 
- Hide all
- Collapse all 
- Expand all
- [ ] Exclusive visibility (toggle- checkbox) 
--- 
- Add sub-group 
- Zoom to group extent (selected in this case may be unclear).
- Remove group
```


# Layer right click menu

- Here we have a lot of tools, and we need to organize them properly. In particular, the only tool that needs to be on toolbar is the swipe tool for 3 reasons: 
    - It should be easy to activate/deactivate, having it in the menu makes deactivation hard.
    - It should be easy to switch between layers. In the old interface, you can change the selected layer while the tool is active (anyway one can right click on another layer to use the tool on the new layer, so this is relatevely easy to handle also with this UI).
    - theoretically sooner or later we should could use swipe/compare with more than one layer at the time (right / left layer). 

The other tools can go all in the right-click menu, without causing any problem. 
As an alternative, I should put in the right menu the configure option, if swipe or spy glass is active, and turn of on click on the icon in the TOC, instead of configure.

In general the tools looks good. we may migrate all the tools in the "tree vertical dots" menu, if present, instead of having them sparse in toolbar and menu.

## Layers

### 
- Visibility of the layers can not be turned off. This makes erroring layers load forever. Maybe aa user want to disable them. I should put the "!" on the right, as for the group. 

[screencast-mail.google.com-2024.03.14-12_21_37.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/e2d92343-1e51-461c-91aa-acdc0c86647d)

File: web/client/hooks/usePluginItems.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, { useRef, useMemo } from 'react';
+import join from 'lodash/join';
+import { getConfiguredPlugin } from '../utils/PluginsUtils';
+
+/**
+ * hook to load and configure items in plugins container
```
Comment (offtherailz) [inline web/client/hooks/usePluginItems.js:14]: can you explain better what this does. 
It looks to apply to item `Component` some attributes spreading config, adding pluginCfg as object, the usage is not clear.

File: web/client/plugins/TOC/index.js [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+ *          TOC: {
+ *              name: "TOOLNAME", // a name for the current tool.
+ *              target: "node", // the target where to insert the component
+ *              Component: MyNodeComponent,
+ *              // selector is `mandatory` and it and it will receive same prop as the Component
+ *              // this example will modify the node of groups
+ *              selector: ({ nodeType, nodeTypes }) => nodeType === nodeTypes.GROUP
+ *          },
+ * // ...
+ * ```
```
Comment (offtherailz) [inline web/client/plugins/TOC/index.js:217]: This explainaition is great for developers, but less useful for configuring the toc. 
I should put it in the `@prop items`. What you think ?  Could this be a best practice for future documentation of this property ?

**Comment by tdipisa** [comment]
Thank you @offtherailz 
I've marked with [X] points that in my opinion should be done or checked now before merging the PR. I've already spoke with @allyoucanmapand he will work on them shortly. For the rest, we will open a dedicated issue to evaluate further improvements after we have all dedicated time to testing/using the component.

**Review by offtherailz — CHANGES_REQUESTED** [review]
During my testing I had this error while dragging and dropping, with a error screen:

[screencast-mail.google.com-2024.03.21-16_23_26.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/3a5142fe-6ee0-48b5-a488-d047114592a8)


```txt
TypeError: Cannot read properties of undefined (reading 'id')
    at eval (layers.js:76:24)
    at Array.map (<anonymous>)
    at layers (layers.js:75:46)
    at standardRootReducerFunc (defaultOptions.js:86:81)
    at rootReducer (StandardStore.js:90:12)
    at Object.dispatch (createStore.js:173:22)
    at dispatch (index.js:17:13)
    at eval (middleware.js:27:18)
    at eval (createEpicMiddleware.js:79:22)
    at eval (createEpicMiddleware.js:79:22)
    at eval (index.js:12:74)
    at Object.eval [as onLayerLoad] (bindActionCreators.js:7:12)
    at Object.next (Layer.jsx:191:26)
    at SafeSubscriber.__tryOrUnsub (Subscriber.js:234:16)
    at SafeSubscriber.next (Subscriber.js:183:22)
    at Subscriber._next (Subscriber.js:125:26)
```
It refers to the layers reducer `LAYER_LOAD` case

```
const newLayers = (state.flat || []).map(layer => {
          return layer.id === action.layerId ?  // <=== HERE
```
I think layer may be null in some cases, but I let you debug and find the proper solution.
The rest looks good.

File: docs/developer-guide/mapstore-migration-guide.md [hunk 3]
```diff
@@ -22,6 +22,86 @@ This is a list of things to check if you want to update from a previous version
...
+    ]
+}
+```
+
+Note: contexts using this tool needs to be updated referring to the configuration of the new MetadataInfo plugin.
+
+### Removal of deprecated FloatingLegend plugin
+
+The Floating legend plugin has been completely removed from the core repository.
+
```
Comment (offtherailz) [inline docs/developer-guide/mapstore-migration-guide.md:104]:
```suggestion
### TOC plugin refactor

The table of content (TOC) has been refactored with following changes:

- Removal of hardcoded tools from the toolbar. Now all the toolbar's elements are injected by related plugins. This includes also the introduction of a new plugin called `MetadataInfo` for layer metadata.
- Refactors of TOC components
- Review of plugin configuration (cfg). List of changed configuration properties:
  - `activateSettingsTool` removed property, now the button will be added directly from `TOCItemsSettings` when available
  - `activateQueryTool` removed property, now the button will be directly added by `FilterLayer` plugin, when available
  - `activateDownloadTool` removed property, now the button will be added directly from `LayerDownload` when available
  - `activateMetedataTool` removed property, now the button will be added directly from `MetadataInfo` when availables
  - `checkPlugins` remove property, now availability of tools rely on the related plugin so this check is not needed anymore  
  - `showFullTitleOnExpand`  removed property, the new style allows for seeing the full title inline without duplicating it
  - `metadataTemplate` this configuration has been moved to `MetadataInfo` plugin
  - `metadataOptions` this configuration has been moved to `MetadataInfo` plugin

### Introduction of MetadataInfo plugin

The `MetadataInfo` plugin has been introduced to include the layer metadata info button of the TOC toolbar has separated plugin as expected by the new TOC.
Some steps are needed to correctly configure it:

- Ensure to import the `MetadataInfo` plugin in the downstream project
- Include the plugin definition in the `pluginsConfig.json` to make it available inside contexts
- Move the related configuration (cfg) from the TOC to the `MetadataInfo` plugin definition in `localConfig.json` file. This is only necessary if the layer metadata button was configured at the application level
- Update your contexts to use `MetadataInfo` plugin, properly configured, when needed

expected changes in the `pluginsConfig.json` file:

- add `MetadataInfo` entry to the list

```js
{
    "name": "MetadataInfo"
}
```

- add `MetadataInfo` to `TOC` children

```js
{
    "name": "TOC",
    "glyph": "1-layer",
    "symbol": "layers",
    "title": "plugins.TOC.title",
    "description": "plugins.TOC.description",
    "defaultConfig": {
        "activateAddLayerButton": true,
        "addLayersPermissions": true,
        "removeLayersPermissions": true,
        "sortingPermissions": true,
        "addGroupsPermissions": true,
        "removeGroupsPermissions": true,
        "activateWidgetTool": true,
        "activateMetedataTool": false
    },
    "children": [
        "TOCItemsSettings",
        "FeatureEditor",
        "FilterLayer",
        "AddGroup",
        "Swipe",
        "MetadataInfo"
    ],
    "autoEnableChildren": [
        "TOCItemsSettings",
        "FeatureEditor",
        "FilterLayer",
        "AddGroup"
    ],
    "dependencies": [
        "DrawerMenu"
    ]
}
```

Note: contexts using this tool needs to be updated referring to the configuration of the new MetadataInfo plugin.

### Removal of deprecated FloatingLegend plugin

The `FloatingLegend` plugin has been completely removed from the core repository.

```

File: web/client/plugins/TOC/index.js [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+ * @prop {boolean} cfg.activateAddLayerButton: activate a button to open the catalog, default `true`
+ * @prop {boolean} cfg.activateAddGroupButton: activate a button to add a new group, default `true`
+ * @prop {boolean} cfg.activateLayerInfoTool: activate a button to enable the layer info
+ * @prop {boolean} [addLayersPermissions=true] if false, only users of role ADMIN can see the "add layers" button. Default true.
+ * @prop {boolean} [removeLayersPermissions=true] if false, only users of role ADMIN have the permission to remove layers. Default true.
+ * @prop {boolean} [sortingPermissions=true] if false, only users of role ADMIN have the permission to move layers in the TOC. Default true.
+ * @prop {boolean} [addGroupsPermissions=true] if false, only users of role ADMIN have the permission to add groups to the TOC. Default true.
+ * @prop {boolean} [removeGroupsPermissions=true] if false, only users of role ADMIN can remove groups from the TOC. Default true.
+ * @prop {boolean} [layerInfoToolPermissions=false] if false, only users of role ADMIN can see the layer info tool. Default false.
```
Comment (offtherailz) [inline web/client/plugins/TOC/index.js:73]:
```suggestion
 * @prop {boolean} [cfg.addLayersPermissions=true] if false, only users of role ADMIN can see the "add layers" button. Default true.
 * @prop {boolean} [cfg.removeLayersPermissions=true] if false, only users of role ADMIN have the permission to remove layers. Default true.
 * @prop {boolean} [cfg.sortingPermissions=true] if false, only users of role ADMIN have the permission to move layers in the TOC. Default true.
 * @prop {boolean} [cfg.addGroupsPermissions=true] if false, only users of role ADMIN have the permission to add groups to the TOC. Default true.
 * @prop {boolean} [cfg.removeGroupsPermissions=true] if false, only users of role ADMIN can remove groups from the TOC. Default true.
 * @prop {boolean} [cfg.layerInfoToolPermissions=false] if false, only users of role ADMIN can see the layer info tool. Default false.
```

**Review by offtherailz — APPROVED** [review]

**Comment by tdipisa** [comment]
@ElenaGallo please test reviewing also test scenarios and let us know if we can backport to 2024.01.xx

**Comment by ElenaGallo** [comment]
@allyoucanmap 

1_ In some languages (such as IT), the TOC settings popup does not display well



2_ When a map is share in embedded mood (use this: https://dev-mapstore.geosolutionsgroup.com/mapstore/embedded.html?forceDrawer=true#/49266)

- Clicking on a layer a Zoom button appears on the toolbar and click-right options report only the zoom button
- CLicking on a layer group a Zoom and Delete buttons appears on the toolbar
- For the Annotation there is not click-right options

https://github.com/geosolutions-it/MapStore2/assets/56537133/60d439ca-9a3a-4276-9b74-ceaae84b7342


3_ The Tooltip/Placement option in General layer Settings should not report the Title option. Now if it is set on Title the tooltip is not available.

https://github.com/geosolutions-it/MapStore2/assets/56537133/96e3d195-08b0-4ee0-bd96-07d6c07875ab


4_ On 3D the Swipe and Spy glass should not be available

**Comment by allyoucanmap** [comment]
> 1_ In some languages (such as IT), the TOC settings popup does not display well

- [ ] I'm working on it

> 2_ When a map is share in embedded mood (use this: https://dev-mapstore.geosolutionsgroup.com/mapstore/embedded.html?forceDrawer=true#/49266)

> Clicking on a layer a Zoom button appears on the toolbar and click-right options report only the zoom button
> CLicking on a layer group a Zoom and Delete buttons appears on the toolbar

This depends on localConfig.json configuration, it seems embedded configuration has `activateRemoveLayer` set to false while `acitvateRemoveGroup` is using the default true value.
Other buttons or menu itemes in the menu/toolbar depends on the configured plugin, eg. Swipe will be shown only if configured as plugin and in this case embedded does not contain this plugin.

@tdipisa should I set also `acitvateRemoveGroup` to false in embedded mode?

image

> For the Annotation there is not click-right options

This is a problem related to the annotation that does not set the bounding box (you can check also on QA, there is the same problem). It seems a regression introduced with the new annotation.

@tdipisa should I solve this while working on this task.

> 3_ The Tooltip/Placement option in General layer Settings should not report the Title option. Now if it is set on Title the tooltip is not available.

We introduced a new behaviour where the tooltip is only visible if different from the content of the title and only when the title is not completely visible.

> 4_ On 3D the Swipe and Spy glass should not be available

- [ ] I'm working on it

> 5_ Right-clicking on the TOC panel and selecting Add subgroup to selected group, the group is added under all groups in the TOC, not as a subgroup of a group.

It seems you are targeting the root node based on the video so the group is added in the correct position. @tdipisa should we do something to make it more clear?

**Comment by tdipisa** [comment]
> > 1_ In some languages (such as IT), the TOC settings popup does not display well
> 
> * [ ]  I'm working on it

Ok, thank you.

> > 2_ When a map is share in embedded mood (use this: https://dev-mapstore.geosolutionsgroup.com/mapstore/embedded.html?forceDrawer=true#/49266)
> 
> > Clicking on a layer a Zoom button appears on the toolbar and click-right options report only the zoom button
> > CLicking on a layer group a Zoom and Delete buttons appears on the toolbar
> 
> This depends on localConfig.json configuration, it seems embedded configuration has `activateRemoveLayer` set to false while `acitvateRemoveGroup` is using the default true value. Other buttons or menu itemes in the menu/toolbar depends on the configured plugin, eg. Swipe will be shown only if configured as plugin and in this case embedded does not contain this plugin.
> 
> @tdipisa should I set also `acitvateRemoveGroup` to false in embedded mode?

Yes please.

> > For the Annotation there is not click-right options
> 
> This is a problem related to the annotation that does not set the bounding box (you can check also on QA, there is the same problem). It seems a regression introduced with the new annotation.
> 
> @tdipisa should I solve this while working on this task.

No, it is better to open a dedicated issue still for 2024.01.00 if it is a regression. @ElenaGallo please create it and assign to @allyoucanmap 

> > 3_ The Tooltip/Placement option in General layer Settings should not report the Title option. Now if it is set on Title the tooltip is not available.
> 
> We introduced a new behaviour where the tooltip is only visible if different from the content of the title and only when the title is not completely visible.

That's fine for now. Please @ElenaGallo open anyway an issue to add helpers on the settings UI to improve the UX and assign it to me

> > 4_ On 3D the Swipe and Spy glass should not be available
> 
> * [ ]  I'm working on it

Ok, thank you.

> > 5_ Right-clicking on the TOC panel and selecting Add subgroup to selected group, the group is added under all groups in the TOC, not as a subgroup of a group.
> 
> It seems you are targeting the root node based on the video so the group is added in the correct position. @tdipisa should we do something to make it more clear?

No, that's fine as it is now I think it is already enough clear, the selected node is highlighted and the same is valid for the root one.

**Comment by ElenaGallo** [comment]
New issues are here: https://github.com/geosolutions-it/MapStore2/issues/10125 ,  https://github.com/geosolutions-it/MapStore2/issues/10126

**Comment by ElenaGallo** [comment]
@allyoucanmap please backport to 2024.01.xx. thanks

**Pull request merged**